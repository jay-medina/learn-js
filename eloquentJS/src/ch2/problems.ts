export function loopingTriangle(length: number, triangleChar = '#'): string {
  if (length <= 0) return '';

  let finalStr = '\n';

  for (let i = 1; i <= length; i += 1) {
    let tempStr = '';
    for (let j = 0; j < i; j += 1) {
      tempStr = tempStr + triangleChar;
    }
    finalStr = finalStr + tempStr + '\n';
  }

  return finalStr;
}

export function fizzbuzz(num: number) {
  let start = 1;
  let arr: string[] = [];

  while(start <= num) {
    if(start % 15 === 0) {
      arr.push('FizzBuzz');
    } else if(start % 5 === 0) {
      arr.push('Buzz')
    } else if(start % 3 === 0) {
      arr.push('Fizz')
    } else {
      arr.push(`${start}`);
    }

    start += 1;
  }

  return arr;
}

export function chessBoard(size: number, tile='#'): string {
  let board = '\n';

  for(let i = 1; i <= size; i += 1) {
    let line = '';
    for(let j = 1; j <= size; j +=1) {
      if((i + j) % 2 === 1) {
        line = line + tile;
      } else {
        line = line + ' ';
      }
    }
    board = board + line + '\n';
  }

  return board;
}
