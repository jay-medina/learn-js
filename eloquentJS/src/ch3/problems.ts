export const min = (a: number, b: number) => {
  if (a < b) return a;

  return b;
};

export const isEven = (a: number): boolean => {
  function check(a: number): boolean {
    if (a === 0) return true;
    if (a === 1) return false;

    return check(a - 2);
  }

  if (a < 0) return check(a * -1);

  return check(a);
};


export const countBs = (str: string): number => countChar(str, 'B');

export const countChar = (str: string, ch: string): number => {
  let ct = 0;

  for(let s of str) {
    if(s === ch) {
      ct += 1;
    }
  }

  return ct;
}
