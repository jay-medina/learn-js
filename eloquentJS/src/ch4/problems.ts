export function range(start: number, end: number, step = 1): number[] {
  if (start > end && step > 0) return [];
  if (start < end && step < 0) return [];

  let arr: number[] = [];

  if (start < end) {
    for (let i = start; i <= end; i = i + step) {
      arr.push(i);
    }
  } else {
    for (let i = start; i >= end; i = i + step) {
      arr.push(i);
    }
  }

  return arr;
}

export function sum(n: number[]): number {
  let s = 0;

  for (let x of n) {
    s += x;
  }

  return s;
}

export function reverseArray<T>(arr: T[]): T[] {
  let reversed: T[] = Array(arr.length);

  for (let i = arr.length - 1; i >= 0; i -= 1) {
    reversed[arr.length - i - 1] = arr[i];
  }

  return reversed;
}

interface Node<T> {
  value: T;
  rest: Node<T> | null;
}

type List<T> = Node<T> | null;

export function arrayToList<T>(arr: T[]): List<T> {
  if (arr.length === 0) return null;

  let li = null;
  let arrReversed = reverseArray(arr);

  for (let a of arrReversed) {
    li = prepend(a, li);
  }

  return li;
}

export function listToArray<T>(list: List<T>): T[] {
  if (!list) return [];

  let arr: T[] = [];
  let pointer: List<T> = list;

  while(pointer != null) {
    arr.push(pointer.value);
    pointer = pointer.rest;
  }

  return arr;
}

export function prepend<T>(element: T, li: List<T>): List<T> {
  return {
    value: element,
    rest: li,
  };
}

// export function nth<T>(li: List<T>, index: number): T | undefined {

// }
