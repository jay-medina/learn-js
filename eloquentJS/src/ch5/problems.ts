import SCRIPTS from './scripts';

export function flatten<T>(arr: T[][]): T[] {
  return arr.reduce((total, next) => {
    return total.concat(next);
  });
}

type Loop = <T>(init: T, condition: (step: T) => boolean, increment: (step: T) => T, body: (step: T) => void) => void;

export const loop: Loop = (init, condition, incre, body) => {
  let step = init;

  while (condition(step)) {
    body(step);
    step = incre(step);
  }
};

export const every = <T>(arr: T[], predicate: (step: T) => boolean): boolean => {
  for (let i of arr) {
    if (!predicate(i)) return false;
  }

  return true;
};

export const everyBySome = <T>(arr: T[], predicate: (step: T) => boolean): boolean => {
  return !some(arr, (n) => !predicate(n));
};

export const some = <T>(arr: T[], predicate: (step: T) => boolean): boolean => {
  for (let i of arr) {
    if (predicate(i)) return true;
  }

  return false;
};

type Direction = 'ltr' | 'rtl' | 'ttb';

export function dominantDirection(text: string): Direction {
  let s = countBy(text.split(''), (ch) => {
    let script = characterScript(ch.codePointAt(0));
    return script ? script.direction : 'none';
  }).filter(({ name }) => name !== 'none');

  return s.reduce((highest, next) => {
    if (highest.count < next.count) return next;

    return highest;
  }).name as Direction;
}

function countBy<T, T2>(items: T[], groupBy: (item: T) => T2) {
  let counts = [];

  for (let item of items) {
    let name = groupBy(item);
    let known = counts.findIndex((c) => c.name == name);
    if (known == -1) {
      counts.push({ name, count: 1 });
    } else {
      counts[known].count++;
    }
  }
  return counts;
}

function characterScript(code: number | undefined) {
  if (!code) return null;

  for (let script of SCRIPTS) {
    if (
      script.ranges.some(([from, to]) => {
        return code >= from && code < to;
      })
    ) {
      return script;
    }
  }
  return null;
}
