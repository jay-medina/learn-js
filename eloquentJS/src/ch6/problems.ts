export class Vec {
    x: number;
    y: number;

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    plus(vec2: Vec) {
        return new Vec(this.x + vec2.x, this.y + vec2.y);
    }

    minus(vec2: Vec) {
        return new Vec(this.x - vec2.x, this.y - vec2.y);
    }

    get length() {
        const x2 = this.x * this.x;
        const y2 = this.y * this.y;
        return Math.sqrt(x2 + y2);
    }
}

function groupIterator<T>(items: T[]): Iterator<T> {
    let currentIndex = 0;
    const { length } = items;

    return {
        next(): IteratorResult<T> {
            if (currentIndex === length) return { done: true, value: undefined! };

            const val = items[currentIndex];
            currentIndex = currentIndex + 1;

            return { value: val, done: false };
        },
    };
}

export class Group<T> implements Iterable<T> {
    private items: T[];

    constructor() {
        this.items = [];
    }

    add(item: T) {
        if (!this.items.includes(item)) {
            this.items.push(item);
        }
    }

    remove(item: T) {
        this.items = this.items.filter(i => i !== item);
    }

    has(item: T): boolean {
        return this.items.includes(item);
    }

    get length() {
        return this.items.length;
    }

    [Symbol.iterator](): Iterator<T> {
        return groupIterator(this.items);
    }

    static from<X>(arr: X[]): Group<X> {
        const g = new Group<X>();
        arr.forEach(i => g.add(i));
        return g;
    }
}
