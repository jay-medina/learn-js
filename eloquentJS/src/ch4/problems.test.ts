import { range, sum, reverseArray, prepend, arrayToList, listToArray, nth, deepEqual } from './problems';

describe('range', () => {
  it('returns the range from start to end', () => {
    expect(range(1, 10)).toMatchSnapshot();
  });

  it('returns the range from 3 to 7', () => {
    expect(range(3, 7)).toMatchSnapshot();
  });

  it('returns an empty array when start > end', () => {
    expect(range(10, 1)).toEqual([]);
  });

  it('returns the range with a step', () => {
    expect(range(1, 10, 2)).toMatchSnapshot();
  });

  it('works with negative steps', () => {
    expect(range(5, 2, -1)).toEqual([5, 4, 3, 2]);
  });

  it('returns empty array when range is negative but start < end', () => {
    expect(range(2, 5, -1)).toEqual([]);
  });
});

describe('sum', () => {
  it('returns the sum of array of numbers', () => {
    expect(sum(range(1, 10))).toBe(55);
  });
});

describe('reverseArray', () => {
  it('returns the array of letters in reverse order', () => {
    expect(reverseArray(['A', 'B', 'C'])).toEqual(['C', 'B', 'A']);
  });

  it('returns the array of numbers in reverse order', () => {
    expect(reverseArray([1, 2, 3, 4, 5])).toEqual([5, 4, 3, 2, 1]);
  });
});

describe('arrayToList', () => {
  it('converts an array of numbers to a linked list', () => {
    expect(arrayToList([1, 2, 3, 4])).toMatchSnapshot();
  });

  it('converts an empty array to null', () => {
    expect(arrayToList([])).toBe(null);
  });
});

describe('listToArray', () => {
  it('converts a list to an array of numbers', () => {
    const original = [1, 2, 3, 4];
    const li = arrayToList(original);

    expect(listToArray(li)).toEqual(original);
  });
});

describe('prepend', () => {
  it('prepends the new element onto an empty list', () => {
    const li = null;

    expect(prepend(10, li)).toMatchSnapshot();
  });

  it('prepends the new element onto the list', () => {
    const li = arrayToList([1, 2, 3, 4]);

    expect(prepend(10, li)).toMatchSnapshot();
  });
});

describe('nth', () => {
  it('returns the nth element', () => {
    const li = arrayToList([1, 2, 3, 4]);

    expect(nth(li, 1)).toBe(2);
  });

  it('returns undefined', () => {
    const li = arrayToList([1, 2, 3, 4]);

    expect(nth(li, 10)).toBeUndefined();
    expect(nth(li, -10)).toBeUndefined();
  });
});

describe('deepEqual', () => {
  it('comparing same object', () => {
    const obj = {here: {is: 'an'}, object: 2};

    expect(deepEqual(obj, obj)).toBe(true);
  });
  it('comparing diff object', () => {
    const obj = {here: {is: 'an'}, object: 2};
    const obj2 = {here: 1, object: 2};

    expect(deepEqual(obj, obj2)).toBe(false);
  });
  it('comparing same object', () => {
    const obj = {here: {is: 'an'}, object: 2};
    const obj2 = {here: {is: 'an'}, object: 2};
    expect(deepEqual(obj, obj2)).toBe(true);
  });
});
