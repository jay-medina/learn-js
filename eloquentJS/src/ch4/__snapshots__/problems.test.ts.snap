// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayToList converts an array of numbers to a linked list 1`] = `
Object {
  "rest": Object {
    "rest": Object {
      "rest": Object {
        "rest": null,
        "value": 4,
      },
      "value": 3,
    },
    "value": 2,
  },
  "value": 1,
}
`;

exports[`prepend prepends the new element onto an empty list 1`] = `
Object {
  "rest": null,
  "value": 10,
}
`;

exports[`prepend prepends the new element onto the list 1`] = `
Object {
  "rest": Object {
    "rest": Object {
      "rest": Object {
        "rest": Object {
          "rest": null,
          "value": 4,
        },
        "value": 3,
      },
      "value": 2,
    },
    "value": 1,
  },
  "value": 10,
}
`;

exports[`range returns the range from 3 to 7 1`] = `
Array [
  3,
  4,
  5,
  6,
  7,
]
`;

exports[`range returns the range from start to end 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;

exports[`range returns the range with a step 1`] = `
Array [
  1,
  3,
  5,
  7,
  9,
]
`;
